webui/react/src/components/UPlot/UPlotChart.tsx

```diff
@@ -26,141 +25,208 @@ interface ScaleZoomData {
-                handler(e);
-              };
-            },
-            mousedown: (_uPlot: uPlot, _target: EventTarget, handler: (e: Event) => void) => {
-              return (e: MouseEvent) => {
-                mousePosition.current = [ e.clientX, e.clientY ];
-                handler(e);
-              };
-            },
-            mouseup: (_uPlot: uPlot, _target: EventTarget, handler: (e: Event) => void) => {
-              return (e: MouseEvent) => {
-                if (!mousePosition.current) {
-                  handler(e);
-                  return;
-                }
-                if (distance(
-                  e.clientX,
-                  e.clientY,
-                  mousePosition.current[0],
-                  mousePosition.current[1],
-                ) > 5) {
-                  isZoomed.current = true;
-                }
-                mousePosition.current = undefined;
-                handler(e);
+              scalesZoomData.current[scaleKey] = {
+                isZoomed,
+                max: maxMax,
+                min: minMin,
               };
             },
-          },
-          drag: { dist: 5, uni: 10, x: true, y: true },
+          ],
         },
-        hooks: {
-          ready: [ (chart: uPlot) => {
-            chartRef.current = chart;
-          } ],
-          setScale: [ (uPlot: uPlot, scaleKey: string) => {
-            const currentMax = uPlot.posToVal(scaleKey === 'x' ? uPlot.bbox.width : 0, scaleKey);
-            const currentMin = uPlot.posToVal(scaleKey === 'x' ? 0 : uPlot.bbox.height, scaleKey);
-            let max = scalesZoomData.current[scaleKey]?.max;
-            let min = scalesZoomData.current[scaleKey]?.min;
-
-            if (max == null || currentMax > max) max = currentMax;
-            if (min == null || currentMin < min) min = currentMin;
-
-            scalesZoomData.current[scaleKey] = { isZoomed: isZoomed.current, max, min };
-
-            /*
-             * Save the scale info if zoomed in and clear it otherwise.
-             * This info will be used to restore the zoom when remounting
-             * the chart, which can be caused by new series data, chart option
-             * changes, etc.
-             */
-            if (!scalesRef.current) scalesRef.current = {};
-            if (isZoomed.current) {
-              scalesRef.current[scaleKey] = uPlot.scales[scaleKey];
-            } else {
-              delete scalesRef.current[scaleKey];
-            }
-            if (Object.keys(scalesRef.current).length === 0) scalesRef.current = undefined;
-          } ],
-        },
-        scales: scalesRef.current,
-        width: chartDivRef.current.offsetWidth,
+        width: chartDivRef.current?.offsetWidth,
       },
-      options,
+      options || {},
     ) as uPlot.Options;
 
-    const plotChart = new uPlot(optionsExtended, normalizedData, chartDivRef.current);
+  const optionsRef = useRef<uPlot.Options>(getAugmentedOptions(options));
```

Comment:
I think a `useRef` is most appropriate for options.

Mainly because it is the most transparent about what is happening.

We are passing options to the `uPlot` constructor as a ref, and it updates it when user interactions occur. so a lot of time `optionsRef.current` will change without any change of the props.